### ISSUE TITLE
assertion failure in getBuiltModel when using option tlimit 

### ISSUE THREAD

**Issue by yannickmoy** [CVC4/CVC4#3269]
We're using option --tlimit-per when calling CVC4 in SPARK as a way to both bound the effort and make results reproducible. On the VC attached, this leads to an assertion failure with the latest stable release 1.7:

$ cvc4-1.7-x86_64-linux-opt --tlimit-per=5000 pbm.smt2
unknown
(error "Assertion failure.
CVC4::theory::TheoryModel* CVC4::TheoryEngine::getBuiltModel()
/barrett/scratch/preiner/release-1.7/dest/src/cvc4-2019-04-09/src/theory/theory_engine.cpp:903:
  d_inSatMode && options::produceModels()
")

[pbm.zip](https://github.com/CVC4/CVC4/files/3601624/pbm.zip)

Is there a way to avoid this failure? 
That might relate to the inability sometimes to produce relevant models when interrupting CVC4, which is of great importance to us, as we'd rather have no model than an incorrect one.
Thanks

**Comment by martin-cs** [comment]
It looks like a failure during model construction.  I have run into similar issues when trying to construct models after an "unknown" response.

For other CVC4 folks:

```
(set-option :produce-models true)
(set-option :incremental true)
;; ...
(check-sat)
(get-model)
```

The relevant parts of the SMT-LIB 2.6 standard are:

[ General semantics ]
"The solver must respond with an error when given a command not permitted in the
current mode."

[ `(check-sat)` ]
"On reporting sat or unknown the solver should move to sat mode—and then respond to
get-assignment, get-model, and get-value commands provided that the corresponding
enabling option is set to true."

[ `(get-model)` ]
"The next three commands can be issued only when the solver is in sat mode, and provide
information related to the most recent check command."

"The model A is required to satisfy all currently asserted
formulas and current assumptions only if the most recent check command reported sat."

So, as I read it, we should be returning something but it can really be anything.

Also, time limits are not necessarily reproducible.  Have you tried?

`(set-option :reproducible-resource-limit 5000)`

**Comment by yannickmoy** [comment]
@martin-cs thanks for answer. Regarding reproducible limits, you're right in that case we're just using the time limit, as this is the call of CVC4 to generate a counterexample, not the call to attempt proof (we generate different VCs for these two calls).

**Comment by ajreynol** [comment]
I can reproduce the assertion failure.

When CVC4 is interrupted, there is no information available regarding the model.  This is different from other unknown responses (e.g. where everything except quantifiers are known to the satisfied).

When you query for a model in this case, I assume you would prefer some kind of recoverable exception saying that the model is not available?  The other option would be to build a 100% arbitrary model, but I assume the former is preferred.

**Comment by yannickmoy** [comment]
@ajreynol definitely the former is better! In our case what we want is simply not to return a model in that case, but if you raise an exception we can catch it and return.

**Comment by ajreynol** [comment]
Fixed in #3286, let us know if you have any issues with the fix.

**Comment by yannickmoy** [comment]
thanks @ajreynol !
